alacritty_terminal/src/window.rs

```diff
@@ -409,6 +422,50 @@ impl Window {
     }
 }
 
+#[cfg(not(any(target_os = "macos", windows)))]
+fn x_embed_window(window: &GlutinWindow, parent_window_id: u64) {
+    let (xlib_display, xlib_window) = match (window.get_xlib_display(), window.get_xlib_window()) {
+        (Some(display), Some(window)) => (display, window),
+        _ => return,
+    };
+
+    let xlib = Xlib::open().expect("get xlib");
+
+    unsafe {
+        let xembed = "_XEMBED".as_ptr();
+        let atom = (xlib.XInternAtom)(xlib_display as *mut _, xembed as *const _, 0);
+        (xlib.XChangeProperty)(
+            xlib_display as _,
+            xlib_window as _,
+            atom,
+            atom,
+            32,
+            PropModeReplace,
+            [0, 1].as_ptr(),
+            2,
+        );
+        
+        // set a handler to deal with unsuccessful reparenting
+        let old_handler = (xlib.XSetErrorHandler)(Some(ctx_error_handler));
+        // reparent the window with a specific handler in-place
+        (xlib.XReparentWindow)(xlib_display as _, xlib_window as _, parent_window_id, 0, 0);
+        // XReparentWindow doesn't give helpful results, so poke the window here
+        let mut attribs = std::mem::MaybeUninit::<x11_dl::xlib::XWindowAttributes>::uninit();
+        (xlib.XGetWindowAttributes)(xlib_display as _, xlib_window as _, attribs.as_mut_ptr());
```

Comment:
I don't quite understand why we're reading the window attributes here?

Reply:
It's sort of explained in the comment.  The result from XReparentWindow isn't a helpful way to detect failures there.  xterm does this the same way.

See the same in their detection here:
https://github.com/joejulian/xterm/blob/defc6dd5684a12dc8e56cb6973ef973e7a32caa3/misc.c#L5744

Reply:
Hmm, that's very interesting. But has lead me to look at how URxvt does it and that might actually be even better.

While URxvt prints more information that indicates they've got an error hadler setup, they basically just check the return type of `XGetWindowAttributes`. If that works (which it does for URxvt apparently), we can just skip the whole error handler nonsense and check for the return type to determine what's going on.

https://github.com/exg/rxvt-unicode/blob/4b8969cc0eab47944f5964c917fda2b4a9c431d7/src/init.C#L1353-L1365

What both seem to do is also updating the viewport dimensions. But since we actually check the viewport size *after* the window was created, I think we should be able to get away with not doing this.

Reply:
I like it, and it works well with less code.  Thanks.